c/src/lib/libbsp/arm/shared/stm32fxxxx/can/hal-can.c

```diff
@@ -0,0 +1,664 @@
+  sFilterConfig.BankNumber            = MAX_FILTERS;
+  
+  if(HAL_CAN_ConfigFilter(hCanHandle, &sFilterConfig) != HAL_OK) 
+  {
+    return -EINVAL;
+  }
+  return RTEMS_SUCCESSFUL;
+  
+}
+
+int stm32_can_init
+(
+    can_bus * self,
+    long baudRate
+)
+{
+
+  stm32_can_bus * bus = (stm32_can_bus * ) self;
+  CAN_HandleTypeDef * hCanHandle = (CAN_HandleTypeDef *) &(bus->wrapper);
+  bus->wrapper.bus = bus;
+  CAN_Instance canInstance = bus->instance;
+
+  CAN_Timing_Values s_timeValues;
+
+  /*Initialize the clock and GPIO pins for the CAN driver*/
+  stm32_can_gpio_init(canInstance);
+
+  /*Configure the CAN peripherals*/
+
+  if (CAN_ONE == canInstance)
+    hCanHandle->Instance = CAN1;
+  else
+    hCanHandle->Instance = CAN2;
+
+  hCanHandle->pTxMsg = &bus->TxMessage;
+  hCanHandle->pRxMsg = &bus->RxMessage;
+
+  hCanHandle->Init.TTCM = DISABLE;
+  hCanHandle->Init.ABOM = DISABLE;
+  hCanHandle->Init.AWUM = DISABLE;
+  hCanHandle->Init.NART = DISABLE;
+  hCanHandle->Init.RFLM = DISABLE;
+  hCanHandle->Init.TXFP = DISABLE;
+  hCanHandle->Init.SJW = CAN_SJW_1TQ;
+  hCanHandle->Init.Mode = CAN_MODE_NORMAL;
+
+  s_timeValues = rtems_can_get_timing_values(baudRate);
+
+  const uint32_t CAN_BS1_OFFSET = 16;
+  const uint32_t CAN_BS2_OFFSET = 20;
+  const uint32_t BS1_MAX = 16;
+  const uint32_t BS2_MAX = 8;
+
+  if (   s_timeValues.s1 > BS1_MAX 
+      || s_timeValues.s1 < 1) 
+  {
+    return CAN_ERROR;
+  }
+
+  if (   s_timeValues.s2 > BS2_MAX 
+      || s_timeValues.s2 < 1) 
+  {
+    return CAN_ERROR;
+  }
+    
+  // The STM32 HAL Code expects Init->BS1 and Init->BS2
+  // to contain the appropriate value so they can simply be 
+  // OR'd into CAN_BTR. 
+  //
+  // We subtract 1, since 1 TQ is represented by zero, but 
+  // our timing function returns the actual number of time 
+  // quanta for S1 and S2
+  hCanHandle->Init.BS1 = (s_timeValues.s1 - 1) << CAN_BS1_OFFSET;
+  hCanHandle->Init.BS2 = (s_timeValues.s2 - 1) << CAN_BS2_OFFSET;
+
+  // The Preescaler however, is expected to be passed in
+  // as is, so we don't need to subtract 1 here.
+  //
+  // See stm32f4xx_hal_can.c, line 309
+  hCanHandle->Init.Prescaler = s_timeValues.prescaler; //s_timeValues.prescaler; //s_timeValues.prescaler;
+
+  if (HAL_CAN_Init(hCanHandle) != HAL_OK)
+  {
+    /* Initialization Error */
+    //TODO::
+    return CAN_ERROR;
+  }
+  CAN_FilterConfTypeDef sFilterConfig;
+  sFilterConfig.FilterNumber = 0;
+  sFilterConfig.FilterMode = CAN_FILTERMODE_IDMASK;
+  sFilterConfig.FilterScale = CAN_FILTERSCALE_32BIT;
+  sFilterConfig.FilterIdHigh = 0x0000;
+  sFilterConfig.FilterIdLow = 0x0000;
+  sFilterConfig.FilterMaskIdHigh = 0x0000;
+  sFilterConfig.FilterMaskIdLow = 0x0000;
+  sFilterConfig.FilterFIFOAssignment = 0;
+  sFilterConfig.FilterActivation = ENABLE;
+  sFilterConfig.BankNumber = 14;
+
+  if(HAL_CAN_ConfigFilter(hCanHandle, &sFilterConfig) != HAL_OK) {
+    //TODO fill in this body
```

Comment:
I understand this is not done, but would be nice if there was a description of what body would do instead of just fill in this body.